File: frontend/appflowy_tauri/src/appflowy_app/application/database/sort/sort_service.ts
import {
  DatabaseViewIdPB,
  DatabaseSettingChangesetPB,
} from '@/services/backend';
import {
  DatabaseEventDeleteAllSorts,
  DatabaseEventGetAllSorts, DatabaseEventUpdateDatabaseSetting,
} from '@/services/backend/events/flowy-database2';
import { pbToSort, Sort } from './sort_types';

export async function getAllSorts(viewId: string): Promise<Sort[]> {
  const payload = DatabaseViewIdPB.fromObject({
    value: viewId,
  });

  const result = await DatabaseEventGetAllSorts(payload);

  return result.map(value => value.items.map(pbToSort)).unwrap();
}

export async function insertSort(viewId: string, sort: Omit<Sort, 'id'>): Promise<void> {
  const payload = DatabaseSettingChangesetPB.fromObject({
    view_id: viewId,
    update_sort: {
      view_id: viewId,
      field_id: sort.fieldId,
      condition: sort.condition,
    },
  });

  const result = await DatabaseEventUpdateDatabaseSetting(payload);

  return result.unwrap();
}

export async function updateSort(viewId: string, sort: Sort): Promise<void> {
  const payload = DatabaseSettingChangesetPB.fromObject({
    view_id: viewId,
    update_sort: {
      view_id: viewId,
      sort_id: sort.id,
      field_id: sort.fieldId,
      condition: sort.condition,
    },
  });

  const result = await DatabaseEventUpdateDatabaseSetting(payload);

  return result.unwrap();
}

export async function deleteSort(viewId: string, sort: Sort): Promise<void> {
  const payload = DatabaseSettingChangesetPB.fromObject({
    view_id: viewId,
    delete_sort: {
      view_id: viewId,
      sort_id: sort.id,
    },
  });

  const result = await DatabaseEventUpdateDatabaseSetting(payload);

  return result.unwrap();
}

export async function deleteAllSorts(viewId: string): Promise<void> {
  const payload = DatabaseViewIdPB.fromObject({
    value: viewId,
  });
  const result = await DatabaseEventDeleteAllSorts(payload);

  return result.unwrap();
}

File: frontend/appflowy_flutter/lib/plugins/database/domain/sort_service.dart
import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/log.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/database_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/setting_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/sort_entities.pb.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:appflowy_result/appflowy_result.dart';

class SortBackendService {
  SortBackendService({required this.viewId});

  final String viewId;

  Future<FlowyResult<List<SortPB>, FlowyError>> getAllSorts() {
    final payload = DatabaseViewIdPB()..value = viewId;

    return DatabaseEventGetAllSorts(payload).send().then((result) {
      return result.fold(
        (repeated) => FlowyResult.success(repeated.items),
        (r) => FlowyResult.failure(r),
      );
    });
  }

  Future<FlowyResult<void, FlowyError>> updateSort({
    required String sortId,
    required String fieldId,
    required SortConditionPB condition,
  }) {
    final insertSortPayload = UpdateSortPayloadPB.create()
      ..viewId = viewId
      ..sortId = sortId
      ..fieldId = fieldId
      ..condition = condition;

    final payload = DatabaseSettingChangesetPB.create()
      ..viewId = viewId
      ..updateSort = insertSortPayload;
    return DatabaseEventUpdateDatabaseSetting(payload).send().then((result) {
      return result.fold(
        (l) => FlowyResult.success(l),
        (err) {
          Log.error(err);
          return FlowyResult.failure(err);
        },
      );
    });
  }

  Future<FlowyResult<void, FlowyError>> insertSort({
    required String fieldId,
    required SortConditionPB condition,
  }) {
    final insertSortPayload = UpdateSortPayloadPB.create()
      ..fieldId = fieldId
      ..viewId = viewId
      ..condition = condition;

    final payload = DatabaseSettingChangesetPB.create()
      ..viewId = viewId
      ..updateSort = insertSortPayload;
    return DatabaseEventUpdateDatabaseSetting(payload).send().then((result) {
      return result.fold(
        (l) => FlowyResult.success(l),
        (err) {
          Log.error(err);
          return FlowyResult.failure(err);
        },
      );
    });
  }

  Future<FlowyResult<void, FlowyError>> reorderSort({
    required String fromSortId,
    required String toSortId,
  }) {
    final payload = DatabaseSettingChangesetPB()
      ..viewId = viewId
      ..reorderSort = (ReorderSortPayloadPB()
        ..viewId = viewId
        ..fromSortId = fromSortId
        ..toSortId = toSortId);

    return DatabaseEventUpdateDatabaseSetting(payload).send();
  }

  Future<FlowyResult<void, FlowyError>> deleteSort({
    required String sortId,
  }) {
    final deleteSortPayload = DeleteSortPayloadPB.create()
      ..sortId = sortId
      ..viewId = viewId;

    final payload = DatabaseSettingChangesetPB.create()
      ..viewId = viewId
      ..deleteSort = deleteSortPayload;

    return DatabaseEventUpdateDatabaseSetting(payload).send().then((result) {
      return result.fold(
        (l) => FlowyResult.success(l),
        (err) {
          Log.error(err);
          return FlowyResult.failure(err);
        },
      );
    });
  }

  Future<FlowyResult<void, FlowyError>> deleteAllSorts() {
    final payload = DatabaseViewIdPB(value: viewId);
    return DatabaseEventDeleteAllSorts(payload).send().then((result) {
      return result.fold(
        (l) => FlowyResult.success(l),
        (err) {
          Log.error(err);
          return FlowyResult.failure(err);
        },
      );
    });
  }
}
